### PR TITLE
Move test setup to code

### PR DESCRIPTION
**Pull request by pgressa**
This PR is about migration of k8s test setup into code in order to have ability to create specific k8s configuration. Where possible the created `micronaut-kubernetes` namespace is reused like it was before. This is basically driven by `spec.reuseNamespace` property that is set to `true` by default. In case of `false` values, the namespace is deleted and all resources are created from the scratch. 

Additionally to make this work a new class `KubernetesSpecification` is introduces and uses `MicronautTest` annotation. 

Apart to that, lot of Required annotations were removed as it is considered that the base test infrastructure was created/reused by `KubernetesSpecification`.

### CODE REVIEW

**Comment by alvarosanchez** [comment]
Do you know why the tests could be failing?

**Comment by pgressa** [comment]
@alvarosanchez this is wrong: 
```
  16:24:26.102 [core] [default-nioEventLoopGroup-5-4] TRACE i.m.h.client.netty.DefaultHttpClient - HTTP Client Streaming Response Received Chunk (length: 0) for Request: GET http://localhost:8001/api/v1/namespaces/micronaut-kubernetes/secrets?labelSelector=
  16:24:26.102 [core] [default-nioEventLoopGroup-5-4] TRACE i.m.h.client.netty.DefaultHttpClient - Chunk Body
  16:24:26.102 [core] [default-nioEventLoopGroup-5-4] TRACE i.m.h.client.netty.DefaultHttpClient - ----
  16:24:26.102 [core] [default-nioEventLoopGroup-5-4] TRACE i.m.h.client.netty.DefaultHttpClient - 
  16:24:26.103 [core] [default-nioEventLoopGroup-5-4] TRACE i.m.h.client.netty.DefaultHttpClient - ----
```

It should return at least the json with empty items array.

**Comment by pgressa** [comment]
Yesterday I managed to successfully run those tests two times in a row, but it was around midnight so I gave up. Now I have rerun the tests and they failed. What I’m thinking of is that the kube-api docker container is basically down, being restarted due to most probably resource needs. It’s hypothesis but :
- based on previous comment when the api did return empty response, it could mean the request was handled by kube-api but it was not properly started yet, hard to say. 
- also it could answer a question why the retry plugin made the tests pass, b/c simply, the kube-api was up and running

So I'll look into two areas now:
- validate that hypothesis
- improve our client b/c at the end of the day, this may happen also in other clusters as well

**Comment by pgressa** [comment]
From [specifications-for-github-hosted-runners#supported-runners-and-hardware-resources](https://docs.github.com/en/free-pro-team@latest/actions/reference/specifications-for-github-hosted-runners#supported-runners-and-hardware-resources):
```
Each virtual machine has the same hardware resources available.

2-core CPU
7 GB of RAM memory
14 GB of SSD disk space
```

Based on K8s installation requirements [install-kubeadm](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/):
```
2 GB or more of RAM per machine (any less will leave little room for your apps)
2 CPUs or more
```

*Basically all containers of K8s control plane have configured `resources.requests`.

**Comment by alvarosanchez** [comment]
Thanks for chasing this. I think you made good points.

We should probably add a configurable retry to all the `KubernetesClient` flowables.

**Comment by pgressa** [comment]
What still bothers me is this kind of response, b/c here the Retry wouldn't be executed since it's HTTP OK:
```
Sending HTTP Request: GET /api/v1/namespaces/micronaut-kubernetes/services?labelSelector=foo%3Dbar
  07:41:00.563 [core] [default-nioEventLoopGroup-16-2] DEBUG i.m.h.client.netty.DefaultHttpClient - Chosen Server: localhost(8001)
  07:41:00.563 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - Accept: application/json
  07:41:00.563 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - host: localhost:8001
  07:41:00.563 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - connection: close
  07:41:00.568 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - HTTP Client Streaming Response Received (200 OK) for Request: GET /api/v1/namespaces/micronaut-kubernetes/services?labelSelector=foo%3Dbar
  07:41:00.568 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - Cache-Control: no-cache, private
  07:41:00.568 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - Content-Length: 151
  07:41:00.573 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - Content-Type: application/json
  07:41:00.573 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - Date: Wed, 02 Dec 2020 07:41:00 GMT
  07:41:00.573 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - Connection: close
  07:41:00.579 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - HTTP Client Streaming Response Received Chunk (length: 151) for Request: GET http://localhost:8001/api/v1/namespaces/micronaut-kubernetes/services?labelSelector=foo%3Dbar
  07:41:00.579 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - Chunk Body
  07:41:00.579 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - ----
  07:41:00.579 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - {"kind":"ServiceList","apiVersion":"v1","metadata":{"selfLink":"/api/v1/namespaces/micronaut-kubernetes/services","resourceVersion":"734"},"items":[]}
  
  07:41:00.579 [core] [default-nioEventLoopGroup-16-2] TRACE i.m.h.client.netty.DefaultHttpClient - ----
```

**Comment by alvarosanchez** [comment]
Right, it would be impossible for us to detect that case. But maybe we can check when the chunk body is empty.

**Comment by pgressa** [comment]
The hypothesis was not confirmed. The k8s control plane is stable during the test. The root cause is that the tests were run in parallel causing the micronaut-test namespace to be deleted and created randomly based on the test configuration. However thanks to the investigation the KubernetesOperations were hardened to be more reliable in terms of creating deloyment and deleting namespaces.

**Comment by alvarosanchez** [comment]
Amazing! Thanks a lot!

Could you please take the opportunity to fix #211 as part of this?

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -20,6 +20,7 @@ jobs:
     strategy:
       matrix:
         java: ['8', '11', '15']
+        k8s: ['1.16', '1.17', '1.18', '1.19']
```
Comment (pgressa) [inline .github/workflows/gradle.yml:23]: this was removed in some previous automatic PR

File: .github/workflows/gradle.yml [hunk 2]
```diff
@@ -33,10 +34,12 @@ jobs:
         with:
           java-version: ${{ matrix.java }}
       - name: Optional setup step
+        env:
+          K8S_VERSION: ${{ matrix.k8s }}        
         run: |
           [ -f ./setup.sh ] && ./setup.sh || true
       - name: Build with Gradle
-        run: ./gradlew dependencyUpdates check --parallel --continue
```
Comment (pgressa) [inline .github/workflows/gradle.yml:39]: parallel won't work here, as majority of the specs are using same namespace. We agreed weeks ago to remove the parallel and in some of the next prs, every spec will create it's own namespace.

File: kubernetes-discovery-client/build.gradle [hunk 3]
```diff
@@ -1,7 +1,3 @@
-plugins {
-    id "org.gradle.test-retry" version "1.2.0"
```
Comment (pgressa) [inline kubernetes-discovery-client/build.gradle:2]: retry is no more needed

File: test-utils/src/main/groovy/io/micronaut/kubernetes/test/KubernetesSpecification.groovy [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+import javax.inject.Inject
+
+/**
+ * Abstract specification encapsulating setup of test namespace.
+ *
+ * @author Pavol Gressa
+ * @since 2.3
+ */
+@Slf4j
+abstract class KubernetesSpecification extends Specification {
```
Comment (pgressa) [inline test-utils/src/main/groovy/io/micronaut/kubernetes/test/KubernetesSpecification.groovy:39]: This is basically the main class that does the setup of resources that was previously in the shell file.

**Comment by pgressa** [comment]
@alvarosanchez this is now ready for review. I will leave the issue with static map for later PRs.

**Pull request merged**